### ISSUE TITLE
ASN.1 compiler fails on circular imports

### ISSUE THREAD

**Issue by joergh** [erlang/otp#4718]
**Describe the bug**
When multiple modules are defined in an ASN.1 specification, exports from later defined modules  are not available. 

**To Reproduce**
Run erlc on the ASN.1 input below, the following errors are shown:

ModuleA:5: 'FieldTypeB' is not exported from ModuleB
ModuleA:7: 'FieldTypeB' is not exported from ModuleB

**Expected behavior**
I would expect it to compile the modules correctly without errors.
 
**Affected versions**
OTP 23.3

**Additional context**
```
ModuleA DEFINITIONS AUTOMATIC TAGS ::= BEGIN

EXPORTS ALL;
IMPORTS FieldTypeB FROM ModuleB;

TypeA ::= SEQUENCE {
        field FieldTypeB
    }
	
FieldTypeA ::= INTEGER(0..1)

END

ModuleB DEFINITIONS AUTOMATIC TAGS ::= BEGIN

EXPORTS ALL;
IMPORTS FieldTypeA FROM ModuleA;

TypeB ::= SEQUENCE {
        field FieldTypeA
    }
	
FieldTypeB ::= INTEGER(0..1)

END
```

**Comment by KennethL** [comment]
I tried it with your files like this:
```
myhost:~/test/asn1$ erlc ModuleA.asn1
myhost:~/test/asn1$ erlc ModuleB.asn1
myhost:~/test/asn1$ ls
ModuleA.asn1    ModuleA.beam  ModuleA.hrl   ModuleB.asn1db  ModuleB.erl
ModuleA.asn1db  ModuleA.erl   ModuleB.asn1  ModuleB.beam    ModuleB.hrl
```
I get no error it seems to work fine.
How do you do to get this to fail?

**Comment by joergh** [comment]
Thank you for your reply!
As for how I can reproduce the problem, I simply put both modules in one file...
That is the most common form of ASN.1 specs that we are getting from third parties - all modules in one large file. It's really not easy to split them into separate files automatically.

**Comment by KennethL** [comment]
Ok, then I understand. The asn1 compiler does not support multiple modules in one file. It has been so from the beginning. A few years ago we got a request to add support for multi module files as that seems to be supported by the standard but we have not been able to prioritize addition of that feature yet. I am not sure it is obvious how that should work with regards to generated files an their names. Seems like all users so far has managed to split the ASN.1 specs into one module per file. Do you have an example of a spec containing multiple modules in one file and where the split is not trivial ?

**Comment by joergh** [comment]
Unfortunately, no, I cannot give you an example ad we are bound by licenses to not supply them to third parties.

As for the single module per file limitation - this is not clear from the documentation at all. Also, an error message to that effect would be nice. It would have spared me some time at least.


Am 9. April 2021 07:39:28 MESZ schrieb Kenneth Lundin ***@***.***>:
>Ok, then I understand. The asn1 compiler does not support multiple
>modules in one file. It has been so from the beginning. A few years ago
>we got a request to add support for multi module files as that seems to
>be supported by the standard but we have not been able to prioritize
>addition of that feature yet. I am not sure it is obvious how that
>should work with regards to generated files an their names. Seems like
>all users so far has managed to split the ASN.1 specs into one module
>per file. Do you have an example of a spec containing multiple modules
>in one file and where the split is not trivial ?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/erlang/otp/issues/4718#issuecomment-816422228

-- 
Jörg Heßdörfer
S.E.A. Datentechnik GmbH
Mülheimer Straße 7
53840 Troisdorf
Tel.: +49 2241 12737-19
Fax.: +49 2241 12737-14
E-Mail: ***@***.***
Web : http://www.sea-gmbh.com

Informationen nach 37aHGB, 35a GmbH-Gesetz:
S.E.A.
Science& Engineering Applications
Datentechnik GmbH
Sitz der Gesellschaft Köln
Handelsregister Köln HRB 27016
Geschäftsführer: Dr. Gerd Schmitz, Wolfram Koerver

**Comment by KennethL** [comment]
I agree that the current behaviour is bad, it either crashes or silently generate code only for the first module in the file, if the first module name matches the file name.
I will take a look and see if it can be improved.
